### PR TITLE
Phase 5c-2a: receiver-side submit_job accept path

### PR DESCRIPTION
**Pull request by bdraco**
# What does this implement/fix?

Phase 5c-2a of [#106](https://github.com/esphome/device-builder/issues/106) — receiver-side **bundle reception + queue + ack**, the half of 5c-2 that drives a submitted bundle from the wire into a queued :class:`FirmwareJob`. Splits the table row's 5c-2 into:

- **5c-2a (this PR)**: bundle reception + ack contract + queue. The "did the job get accepted?" half.
- **5c-2b (next)**: firmware-event fan-out — subscribe to `JOB_*` events filtered to `remote_peer == session.dashboard_id` and push `job_state_changed` / `job_output` frames over the submitting session.

Splitting was a sizing call — even just the receive path is ~560 lines of production + ~17 unit tests; landing it cleanly before the fan-out keeps each PR focused.

**What this PR adds:**

- **`controllers/remote_build_submit_job.py`** — `SubmitJobReceiver` class plus `_PendingSubmit` per-session dataclass. Owns the receive-loop dispatch for `AppMessageType.SUBMIT_JOB` / `SUBMIT_JOB_CHUNK`, drives `BundleAssembler` from 5c-1, writes the assembled tarball + extracts via `esphome.bundle.prepare_bundle_for_compile()` (preserves `.esphome/` / `.pioenvs/` for incremental compiles — load-bearing for the per-peer per-device subtree), and queues a `FirmwareJob` with `remote_peer=session.dashboard_id`.

- **Per-peer per-device subtree** at `<config>/.esphome/.remote_builds/<dashboard_id>/<device_name>/`. Two-segment key dedupes correctly across multi-offloader fleets (two HA Greens both shipping a "kitchen" device land in distinct subtrees) without colliding within one offloader's pool. PlatformIO's `.pio/build` cache then sees stable source paths and skips the cold-rebuild hit.

- **`FirmwareJob.remote_peer`** — new optional string field (default `""`) carrying the offloader's `dashboard_id` when the job came in via peer-link. Empty for locally-submitted jobs. Threads through `FirmwareController._create_job(... remote_peer=...)`. Surfaced as a "from \<peer\>" badge in the Firmware tasks UI later.

- **Receive-loop wiring** in `controllers/remote_build_peer_link._receive_loop` — two new branches dispatch the new frame types to `controller.get_submit_job_receiver()`. `controller` is now threaded into the loop; existing tests pass `MagicMock()` for the controller arg since their inputs don't include the new types.

- **Lifecycle integration** — `RemoteBuildController.start()` constructs the `SubmitJobReceiver` (gated on `self._db.firmware is not None`); `unregister_peer_link_session()` calls `discard_session(dashboard_id)` so a bundle reception that was mid-stream when the session ended doesn't outlive its session.

**Reject reason taxonomy.** Eight distinct `submit_job_ack.reason` codes — six from the 5c-1 `BundleAssemblerErrorCode` (mapped through 1:1) plus two receiver-side codes for paths after assembly (`extract_failed` / `queue_rejected`) and four for header / chunk dispatch (`duplicate_submit` / `invalid_header` / `no_inflight_submit` / `job_id_mismatch` / `chunk_decode_failed`). Wire-level misbehaviour codes (out-of-order chunks, decode failures, post-completion frames) trigger `terminate{malformed_frame}` after the ack; recoverable errors (oversized / undersized / hash mismatch / extract / queue) ack-and-stay so the offloader can retry on a fresh submit.

**17 unit tests** in `tests/test_remote_build_submit_job.py` covering header validation (duplicate / invalid target / oversized), chunk dispatch (no-inflight, job_id mismatch, decode failure, out-of-order, hash mismatch), the happy path through a stubbed `prepare_bundle_for_compile`, extraction failure, and the session-discard contract. End-to-end against the real `FirmwareController` + a real bundle goes in 5c-2b once the firmware-event fan-out lands.

**Related issue or feature (if applicable):**

- esphome/device-builder#106 — phase 5c-2 (split into 5c-2a / 5c-2b).

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue) — `bugfix`
- [x] New feature (non-breaking change which adds functionality) — `new-feature`
- [ ] Enhancement to an existing feature — `enhancement`
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected) — `breaking-change`
- [ ] Refactor (no behaviour change) — `refactor`
- [ ] Documentation only — `docs`
- [ ] Maintenance / chore — `maintenance`
- [ ] CI / workflow change — `ci`
- [ ] Dependencies bump — `dependencies`

## Frontend coordination

- [x] No frontend change needed
- [ ] Companion frontend PR: esphome/device-builder-dashboard-frontend#<number>

(`FirmwareJob.remote_peer` is a new field but it defaults to `""`, so existing frontend renderers ignoring unknown fields keep working unchanged. The "from \<peer\>" badge UI lands in a later phase alongside the offloader Settings UI.)

## Checklist

- [x] The code change is tested and works locally.
- [x] Pre-commit hooks pass (`ruff`, `codespell`, yaml/json/python checks).
- [x] Tests have been added or updated under `tests/` where applicable.
- [x] `components.json` has **not** been hand-edited (regenerate via `script/sync_components.py` if a sync is needed).
- [ ] Architecture-level changes are reflected in `docs/ARCHITECTURE.md` and/or `docs/API.md`.

(Architecture mention deferred to 5c-2b — the receive-then-fan-out shape is what `docs/ARCHITECTURE.md`'s "Long-lived peer-link sessions" section will want to describe end-to-end, not the half landed here.)

### CODE REVIEW

File: esphome_device_builder/controllers/remote_build_peer_link.py [hunk 1]
```diff
@@ -712,6 +717,21 @@ async def _receive_loop(session: PeerLinkSession) -> None:
...
+            # parse said this is a submit_job; treat it as the
+            # matching TypedDict" hand-off — runtime field
+            # validation is the receiver's job.
+            await controller.get_submit_job_receiver().handle_submit_job(
+                session, cast("SubmitJobFrameData", parsed)
+            )
+            continue
+        if msg_type == AppMessageType.SUBMIT_JOB_CHUNK.value:
+            await controller.get_submit_job_receiver().handle_submit_job_chunk(
+                session, cast("SubmitJobChunkFrameData", parsed)
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome_device_builder/controllers/remote_build_peer_link.py]: Fixed in a0728a0. Dropped the string quoting from both `cast()` calls so the `SubmitJobFrameData` / `SubmitJobChunkFrameData` imports are runtime-used (which they already were — ruff just needed the symbols visible at the call site).

File: tests/test_remote_build_submit_job.py [hunk 2]
```diff
@@ -0,0 +1,459 @@
...
+    payload = _ack_payload(session)
+    assert payload["accepted"] is False
+    assert payload["reason"] == "out_of_order"
+    session.terminate.assert_awaited_once()
+
+
+@pytest.mark.asyncio
+async def test_submit_job_chunk_hash_mismatch_recoverable(
+    tmp_path: Path, monkeypatch: pytest.MonkeyPatch
+) -> None:
```
[the remark this answers is not in the document]
Comment (bdraco) [reply tests/test_remote_build_submit_job.py]: Fixed in a0728a0 — dropped the unused `monkeypatch` parameter.

File: esphome_device_builder/controllers/remote_build_submit_job.py [hunk 3]
```diff
@@ -0,0 +1,550 @@
...
+
+        Rejects (with a typed ``submit_job_ack``) on:
+
+        * Duplicate submit while a previous one is still in
+          flight on the same session.
+        * Header field shapes the wire-format TypedDict can't
+          enforce at runtime (target outside the
+          ``compile`` / ``upload`` set, malformed
+          ``configuration_filename``).
+        * Assembler-construction validation (oversized total,
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome_device_builder/controllers/remote_build_submit_job.py:323]: Good catch — fixed in a0728a0. New `_validate_configuration_filename` helper rejects empty input, path separators (`/` / `\\`), NUL bytes, reserved names (`.` / `..`), leading-dot escapes (`..yaml`), and anything not ending in `.yaml` / `.yml`. `handle_submit_job` calls it upfront and rejects with `invalid_header` on failure. 11 new parameterised tests cover the rejection set.

File: esphome_device_builder/controllers/remote_build_submit_job.py [hunk 4]
```diff
@@ -0,0 +1,550 @@
...
+        wire-level misbehaviour).
+
+        Disk I/O hops to the executor:
+        ``prepare_bundle_for_compile`` walks the tar, validates
+        members, writes to disk; bundling that into one
+        ``run_in_executor`` keeps the receiver's WS dispatch
+        coroutine non-blocking through a multi-MB write.
+        """
+        device_name = _device_name_from_filename(pending.configuration_filename)
+        target_dir = self._config_dir / _REMOTE_BUILDS_SUBDIR / session.dashboard_id / device_name
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome_device_builder/controllers/remote_build_submit_job.py]: Good catch — fixed in a0728a0. Two-layer defense: (1) `_validate_configuration_filename` catches separator / `..` shapes in `configuration_filename` upfront and rejects the header. (2) Inside `_extract_and_queue` a `target_dir.resolve().relative_to(<config>/.esphome/.remote_builds)` check confirms the assembled path stays under the remote-builds root — catches anything an exotic `dashboard_id` shape would slip past (today the dashboard_id is captured from the Noise transcript and never validated for path-component safety; the resolve-and-stay-under-root gate covers that). New regression test `test_submit_job_path_traversal_dashboard_id_caught_at_extract` drives a malicious `dashboard_id="../../escape"` and asserts the reject lands as `invalid_header`.

File: esphome_device_builder/controllers/remote_build_submit_job.py [hunk 5]
```diff
@@ -0,0 +1,618 @@
...
+          header values, so they map to a ``submit_job_ack``
+          rejection rather than a ``terminate{malformed_frame}``;
+          the chunk stream hasn't started yet, the wire is still
+          intact.
+        """
+        if session.dashboard_id in self._inflight:
+            await self._reject(session, job_id=frame["job_id"], reason=_REASON_DUPLICATE_SUBMIT)
+            return
+        target = frame["target"]
+        if target not in _TARGET_TO_JOB_TYPE:
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome_device_builder/controllers/remote_build_submit_job.py:353]: Good catch — fixed in 829a2ac. New `_validate_frame_shape(frame, required)` helper walks each contract (`_SUBMIT_JOB_HEADER_FIELDS` / `_SUBMIT_JOB_CHUNK_FIELDS`) checking presence + types before any indexing. `handle_submit_job` calls it upfront and rejects with `invalid_header` + `terminate{malformed_frame}` on a malformed frame. `bool` is special-cased so a frame announcing `total_bundle_bytes=True` does not sneak past the `int` check. 5 new parameterised tests cover missing fields, wrong types, and the bool-as-int edge case.

File: esphome_device_builder/controllers/remote_build_submit_job.py [hunk 6]
```diff
@@ -0,0 +1,618 @@
...
+        whether the failure is wire-level (offloader corrupted
+        the stream — close the session) or recoverable (offloader
+        can retry on a fresh submit). Happy-path completion
+        flows through :meth:`_finalise_and_queue`.
+        """
+        pending = self._inflight.get(session.dashboard_id)
+        if pending is None:
+            await self._reject(session, job_id=frame["job_id"], reason=_REASON_NO_INFLIGHT)
+            return
+        if frame["job_id"] != pending.job_id:
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome_device_builder/controllers/remote_build_submit_job.py:415]: Same shape-validation gate applied to `handle_submit_job_chunk` in 829a2ac. Maps to a new `invalid_chunk` reason (distinct from `invalid_header` so the offloader can tell whether the contract drift was on the header or the stream). Drops the in-flight state + terminates because mid-stream wire corruption is unrecoverable. 4 parameterised tests pin the rejection set.

File: tests/test_remote_build_peer_link.py [hunk 7]
```diff
@@ -59,7 +61,19 @@
...
+    SubmitJobReceiver,
+)
 from esphome_device_builder.helpers import json as _json
+from esphome_device_builder.helpers.peer_link_bundle import (
+    chunk_bundle as _chunk_bundle,
+)
+from esphome_device_builder.helpers.peer_link_bundle import (
+    compute_bundle_sha256 as _compute_bundle_sha256,
+)
+from esphome_device_builder.helpers.peer_link_bundle import (
```
[the remark this answers is not in the document]
Comment (bdraco) [reply tests/test_remote_build_peer_link.py]: Already addressed in a3d59a8 — the bundle-helper imports moved into a single block inside `tests/conftest.py` (the new `make_submit_job_frames` helper). `tests/test_remote_build_peer_link.py` no longer imports from the bundle module at all.

File: esphome_device_builder/controllers/remote_build_submit_job.py [hunk 8]
```diff
@@ -0,0 +1,708 @@
...
+            raise _SubmitJobRejectionError(_REASON_EXTRACT_FAILED) from exc
+
+        # ``configuration`` on FirmwareJob is the path relative to
+        # the controller's config_dir (``rel_path`` joins back on
+        # the way out). The extracted YAML lives under
+        # ``.esphome/.remote_builds/<dashboard_id>/<device_name>/``
+        # inside the same config_dir, so ``relative_to`` always
+        # succeeds here.
+        rel_yaml = extracted_yaml.relative_to(self._config_dir)
+        configuration = str(rel_yaml)
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome_device_builder/controllers/remote_build_submit_job.py]: Good catch — fixed in bd2e53f. Swapped to `rel_yaml.as_posix()` so `FirmwareJob.configuration` carries forward-slash separators regardless of receiver platform. The wire-side payload stays stable across a Linux-receiver / Windows-receiver mix even though the filesystem-level `rel_path` join would resolve either form correctly.
